Generated clients are missing routing headers

https://aip.dev/client-libraries/4222

It looks like the generator is not adding variables declared in the `google.api.http` annotation to routing headers.

https://github.com/googleapis/googleapis/blob/52701da10fec2a5f9796e8d12518c0fe574488fe/google/cloud/servicedirectory/v1beta1/registration_service.proto#L60-L65

See https://github.com/GoogleCloudPlatform/python-docs-samples/pull/3456/files#r412305714 for the modifications that need to be made to currently use the library.

Possibly related to #380.

Metadata should be added [somewhere here](https://github.com/googleapis/python-service-directory/blob/25e0fcfb32d100be0cba3799d62543569dd2d2c6/google/cloud/servicedirectory_v1beta1/services/registration_service/client.py#L359-L378) (see [this method](https://github.com/googleapis/python-service-directory/blob/25e0fcfb32d100be0cba3799d62543569dd2d2c6/google/cloud/servicedirectory_v1beta1/services/registration_service/client.py#L432-L459) for comparison))

Bumping the priority on this as I'm seeing the same issue with KMS. 

Errors look like:

```
E           grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
E               status = StatusCode.NOT_FOUND
E               details = "The request concerns location 'us-east1' but was sent to location 'global'. Either Cloud KMS is not available in 'us-east1' or the request was misrouted. Make sure the 'x-goog-request-params' metadata is set correctly; see https://cloud.google.com/kms/docs/grpc for more information."
E               debug_error_string = "{"created":"@1591243631.459760738","description":"Error received from peer ipv4:74.125.142.95:443","file":"src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"The request concerns location 'us-east1' but was sent to location 'global'. Either Cloud KMS is not available in 'us-east1' or the request was misrouted. Make sure the 'x-goog-request-params' metadata is set correctly; see https://cloud.google.com/kms/docs/grpc for more information.","grpc_status":5}"
E           >
```

--- a/gapic/ads-templates/%namespace/%name/%version/%sub/services/%service/client.py.j2
+++ b/gapic/ads-templates/%namespace/%name/%version/%sub/services/%service/client.py.j2
@@ -117,7 +117,7 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
     from_service_account_json = from_service_account_file
 
 
-    {% for message in service.resource_messages -%}
+    {% for message in service.resource_messages|sort(attribute="resource_type") -%}
     @staticmethod
     def {{ message.resource_type|snake_case }}_path({% for arg in message.resource_path_args %}{{ arg }}: str,{% endfor %}) -> str:
         """Return a fully-qualified {{ message.resource_type|snake_case }} string."""

--- a/gapic/ads-templates/%namespace/%name/%version/%sub/services/%service/client.py.j2
+++ b/gapic/ads-templates/%namespace/%name/%version/%sub/services/%service/client.py.j2
@@ -333,7 +333,9 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
         metadata = tuple(metadata) + (
             gapic_v1.routing_header.to_grpc_metadata((
                 {%- for field_header in method.field_headers %}
+                {%- if not method.client_streaming %}
                 ('{{ field_header }}', request.{{ field_header }}),
+                {%- endif %}
                 {%- endfor %}
             )),
         )

--- a/gapic/ads-templates/tests/unit/%name_%version/%sub/test_%service.py.j2
+++ b/gapic/ads-templates/tests/unit/%name_%version/%sub/test_%service.py.j2
@@ -262,21 +262,29 @@ def test_{{ method.name|snake_case }}(transport: str = 'grpc'):
 def test_{{ method.name|snake_case }}_field_headers():
     client = {{ service.client_name }}(
         credentials=credentials.AnonymousCredentials(),
-  )
+    )
 
     # Any value that is part of the HTTP/1.1 URI should be sent as
     # a field header. Set these to a non-empty value.
-    request = {{ method.input.ident }}(
-        {%- for field_header in method.field_headers %}
-        {{ field_header }}='{{ field_header }}/value',
-        {%- endfor %}
-    )
+    request = {{ method.input.ident }}()
+
+    {%- for field_header in method.field_headers %}
+    request.{{ field_header }} = '{{ field_header }}/value'
+    {%- endfor %}
 
     # Mock the actual call within the gRPC stub, and fake the request.
     with mock.patch.object(
             type(client._transport.{{ method.name|snake_case }}),
             '__call__') as call:
+        {% if method.void -%}
+        call.return_value = None
+        {% elif method.lro -%}
+        call.return_value = operations_pb2.Operation(name='operations/op')
+        {% elif method.server_streaming -%}
+        call.return_value = iter([{{ method.output.ident }}()])
+        {% else -%}
         call.return_value = {{ method.output.ident }}()
+        {% endif %}
         client.{{ method.name|snake_case }}(request)
 
         # Establish that the underlying gRPC stub method was called.
@@ -470,7 +478,6 @@ def test_{{ method.name|snake_case }}_raw_page_lro():
 {% endif %} {#- method.paged_result_field #}
 
 {% endfor -%} {#- method in methods #}
-
 
 def test_credentials_transport_error():
     # It is an error to provide credentials and a transport instance.
@@ -689,12 +696,12 @@ def test_{{ service.name|snake_case }}_grpc_lro_client():
 {% for message in service.resource_messages -%}
 {% with molluscs = cycler("squid", "clam", "whelk", "octopus", "oyster", "nudibranch", "cuttlefish", "mussel", "winkle", "nautilus", "scallop", "abalone") -%}
 def test_{{ message.resource_type|snake_case }}_path():
-  {% for arg in message.resource_path_args -%}
-  {{ arg }} = "{{ molluscs.next() }}"
-  {% endfor %}
-  expected = "{{ message.resource_path }}".format({% for arg in message.resource_path_args %}{{ arg }}={{ arg }}, {% endfor %})
-  actual = {{ service.client_name }}.{{ message.resource_type|snake_case }}_path({{message.resource_path_args|join(", ") }})
-  assert expected == actual
+    {% for arg in message.resource_path_args -%}
+    {{ arg }} = "{{ molluscs.next() }}"
+    {% endfor %}
+    expected = "{{ message.resource_path }}".format({% for arg in message.resource_path_args %}{{ arg }}={{ arg }}, {% endfor %})
+    actual = {{ service.client_name }}.{{ message.resource_type|snake_case }}_path({{message.resource_path_args|join(", ") }})
+    assert expected == actual
 
 
 def test_parse_{{ message.resource_type|snake_case }}_path():

--- a/gapic/ads-templates/tests/unit/%name_%version/%sub/test_%service.py.j2
+++ b/gapic/ads-templates/tests/unit/%name_%version/%sub/test_%service.py.j2
@@ -258,7 +258,7 @@ def test_{{ method.name|snake_case }}(transport: str = 'grpc'):
     {% endfor %}
     {% endif %}
 
-{% if method.field_headers %}
+{% if method.field_headers and not method.client_streaming %}
 def test_{{ method.name|snake_case }}_field_headers():
     client = {{ service.client_name }}(
         credentials=credentials.AnonymousCredentials(),

--- a/gapic/schema/wrappers.py
+++ b/gapic/schema/wrappers.py
@@ -628,20 +628,16 @@ class Method:
 
         pattern = re.compile(r'\{([a-z][\w\d_.]+)=')
 
-        if http.get:
-            return tuple(pattern.findall(http.get))
-        elif http.put:
-            return tuple(pattern.findall(http.put))
-        elif http.post:
-            return tuple(pattern.findall(http.post))
-        elif http.delete:
-            return tuple(pattern.findall(http.delete))
-        elif http.patch:
-            return tuple(pattern.findall(http.patch))
-        elif http.custom.path:
-            return tuple(pattern.findall(http.custom.path))
+        potential_verbs = [
+            http.get,
+            http.put,
+            http.post,
+            http.delete,
+            http.patch,
+            http.custom.path,
+        ]
 
-        return ()
+        return next((tuple(pattern.findall(verb)) for verb in potential_verbs if verb), ())
 
     @utils.cached_property
     def flattened_fields(self) -> Mapping[str, Field]:

--- a/gapic/schema/wrappers.py
+++ b/gapic/schema/wrappers.py
@@ -638,7 +638,7 @@ class Method:
             return tuple(pattern.findall(http.delete))
         elif http.patch:
             return tuple(pattern.findall(http.patch))
-        elif http.custom:
+        elif http.custom.path:
             return tuple(pattern.findall(http.custom.path))
 
         return ()

--- a/gapic/schema/wrappers.py
+++ b/gapic/schema/wrappers.py
@@ -625,8 +625,22 @@ class Method:
     def field_headers(self) -> Sequence[str]:
         """Return the field headers defined for this method."""
         http = self.options.Extensions[annotations_pb2.http]
+
+        pattern = re.compile(r'\{([a-z][\w\d_.]+)=')
+
         if http.get:
-            return tuple(re.findall(r'\{([a-z][\w\d_.]+)=', http.get))
+            return tuple(pattern.findall(http.get))
+        elif http.put:
+            return tuple(pattern.findall(http.put))
+        elif http.post:
+            return tuple(pattern.findall(http.post))
+        elif http.delete:
+            return tuple(pattern.findall(http.delete))
+        elif http.patch:
+            return tuple(pattern.findall(http.patch))
+        elif http.custom:
+            return tuple(pattern.findall(http.custom.path))
+
         return ()
 
     @utils.cached_property

--- a/gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2
+++ b/gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2
@@ -117,7 +117,7 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
     from_service_account_json = from_service_account_file
 
 
-    {% for message in service.resource_messages -%}
+    {% for message in service.resource_messages|sort(attribute="resource_type") -%}
     @staticmethod
     def {{ message.resource_type|snake_case }}_path({% for arg in message.resource_path_args %}{{ arg }}: str,{% endfor %}) -> str:
         """Return a fully-qualified {{ message.resource_type|snake_case }} string."""

--- a/gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2
+++ b/gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2
@@ -333,7 +333,9 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
         metadata = tuple(metadata) + (
             gapic_v1.routing_header.to_grpc_metadata((
                 {%- for field_header in method.field_headers %}
+                {%- if not method.client_streaming %}
                 ('{{ field_header }}', request.{{ field_header }}),
+                {%- endif %}
                 {%- endfor %}
             )),
         )

--- a/gapic/templates/tests/unit/%name_%version/%sub/test_%service.py.j2
+++ b/gapic/templates/tests/unit/%name_%version/%sub/test_%service.py.j2
@@ -258,7 +258,7 @@ def test_{{ method.name|snake_case }}(transport: str = 'grpc'):
     {% endfor %}
     {% endif %}
 
-{% if method.field_headers %}
+{% if method.field_headers and not method.client_streaming %}
 def test_{{ method.name|snake_case }}_field_headers():
     client = {{ service.client_name }}(
         credentials=credentials.AnonymousCredentials(),

--- a/gapic/templates/tests/unit/%name_%version/%sub/test_%service.py.j2
+++ b/gapic/templates/tests/unit/%name_%version/%sub/test_%service.py.j2
@@ -262,21 +262,29 @@ def test_{{ method.name|snake_case }}(transport: str = 'grpc'):
 def test_{{ method.name|snake_case }}_field_headers():
     client = {{ service.client_name }}(
         credentials=credentials.AnonymousCredentials(),
-  )
+    )
 
     # Any value that is part of the HTTP/1.1 URI should be sent as
     # a field header. Set these to a non-empty value.
-    request = {{ method.input.ident }}(
-        {%- for field_header in method.field_headers %}
-        {{ field_header }}='{{ field_header }}/value',
-        {%- endfor %}
-    )
+    request = {{ method.input.ident }}()
+
+    {%- for field_header in method.field_headers %}
+    request.{{ field_header }} = '{{ field_header }}/value'
+    {%- endfor %}
 
     # Mock the actual call within the gRPC stub, and fake the request.
     with mock.patch.object(
             type(client._transport.{{ method.name|snake_case }}),
             '__call__') as call:
+        {% if method.void -%}
+        call.return_value = None
+        {% elif method.lro -%}
+        call.return_value = operations_pb2.Operation(name='operations/op')
+        {% elif method.server_streaming -%}
+        call.return_value = iter([{{ method.output.ident }}()])
+        {% else -%}
         call.return_value = {{ method.output.ident }}()
+        {% endif %}
         client.{{ method.name|snake_case }}(request)
 
         # Establish that the underlying gRPC stub method was called.

--- a/tests/unit/schema/wrappers/test_method.py
+++ b/tests/unit/schema/wrappers/test_method.py
@@ -235,7 +235,8 @@ def test_method_field_headers_present_patch():
 
 
 def test_method_field_headers_present_custom():
-    custom_pattern = http_pb2.CustomHttpPattern(kind="custom", path='/v1/{parent=projects/*}/topics')
+    custom_pattern = http_pb2.CustomHttpPattern(kind='custom',
+        path='/v1/{parent=projects/*}/topics')
     http_rule = http_pb2.HttpRule(custom=custom_pattern)
     method = make_method('DoSomething', http_rule=http_rule)
     assert method.field_headers == ('parent',)

--- a/tests/unit/schema/wrappers/test_method.py
+++ b/tests/unit/schema/wrappers/test_method.py
@@ -204,8 +204,39 @@ def test_method_field_headers_none():
     assert isinstance(method.field_headers, collections.abc.Sequence)
 
 
-def test_method_field_headers_present():
+def test_method_field_headers_present_get():
     http_rule = http_pb2.HttpRule(get='/v1/{parent=projects/*}/topics')
+    method = make_method('DoSomething', http_rule=http_rule)
+    assert method.field_headers == ('parent',)
+
+
+def test_method_field_headers_present_put():
+    http_rule = http_pb2.HttpRule(put='/v1/{parent=projects/*}/topics')
+    method = make_method('DoSomething', http_rule=http_rule)
+    assert method.field_headers == ('parent',)
+
+
+def test_method_field_headers_present_post():
+    http_rule = http_pb2.HttpRule(post='/v1/{parent=projects/*}/topics')
+    method = make_method('DoSomething', http_rule=http_rule)
+    assert method.field_headers == ('parent',)
+
+
+def test_method_field_headers_present_delete():
+    http_rule = http_pb2.HttpRule(delete='/v1/{parent=projects/*}/topics')
+    method = make_method('DoSomething', http_rule=http_rule)
+    assert method.field_headers == ('parent',)
+
+
+def test_method_field_headers_present_patch():
+    http_rule = http_pb2.HttpRule(patch='/v1/{parent=projects/*}/topics')
+    method = make_method('DoSomething', http_rule=http_rule)
+    assert method.field_headers == ('parent',)
+
+
+def test_method_field_headers_present_custom():
+    custom_pattern = http_pb2.CustomHttpPattern(kind="custom", path='/v1/{parent=projects/*}/topics')
+    http_rule = http_pb2.HttpRule(custom=custom_pattern)
     method = make_method('DoSomething', http_rule=http_rule)
     assert method.field_headers == ('parent',)
 

--- a/tests/unit/schema/wrappers/test_method.py
+++ b/tests/unit/schema/wrappers/test_method.py
@@ -204,42 +204,19 @@ def test_method_field_headers_none():
     assert isinstance(method.field_headers, collections.abc.Sequence)
 
 
-def test_method_field_headers_present_get():
-    http_rule = http_pb2.HttpRule(get='/v1/{parent=projects/*}/topics')
-    method = make_method('DoSomething', http_rule=http_rule)
-    assert method.field_headers == ('parent',)
+def test_method_field_headers_present():
+    verbs = [
+        'get',
+        'put',
+        'post',
+        'delete',
+        'patch',
+    ]
 
-
-def test_method_field_headers_present_put():
-    http_rule = http_pb2.HttpRule(put='/v1/{parent=projects/*}/topics')
-    method = make_method('DoSomething', http_rule=http_rule)
-    assert method.field_headers == ('parent',)
-
-
-def test_method_field_headers_present_post():
-    http_rule = http_pb2.HttpRule(post='/v1/{parent=projects/*}/topics')
-    method = make_method('DoSomething', http_rule=http_rule)
-    assert method.field_headers == ('parent',)
-
-
-def test_method_field_headers_present_delete():
-    http_rule = http_pb2.HttpRule(delete='/v1/{parent=projects/*}/topics')
-    method = make_method('DoSomething', http_rule=http_rule)
-    assert method.field_headers == ('parent',)
-
-
-def test_method_field_headers_present_patch():
-    http_rule = http_pb2.HttpRule(patch='/v1/{parent=projects/*}/topics')
-    method = make_method('DoSomething', http_rule=http_rule)
-    assert method.field_headers == ('parent',)
-
-
-def test_method_field_headers_present_custom():
-    custom_pattern = http_pb2.CustomHttpPattern(kind='custom',
-        path='/v1/{parent=projects/*}/topics')
-    http_rule = http_pb2.HttpRule(custom=custom_pattern)
-    method = make_method('DoSomething', http_rule=http_rule)
-    assert method.field_headers == ('parent',)
+    for v in verbs:
+        rule = http_pb2.HttpRule(**{v: '/v1/{parent=projects/*}/topics'})
+        method = make_method('DoSomething', http_rule=rule)
+        assert method.field_headers == ('parent',)
 
 
 def test_method_idempotent_yes():

